### PR TITLE
indexes for mysql schema dump

### PR DESCRIPTION
**Pull request by jwh315**
I ran across your schema dump/load feature for phinx a few weeks ago.  I really love your implementation but really needed indexes.  I was able to get the indexes put into the mysql adapter pretty easily.  

I have a pretty decent size dev database (over 700 tables) at my company, that has an array of ridiculous design inconsistencies from years of development.  The indexing seems to be recreating tables 100% from their original state, so I am fairly confident that I have covered most edge cases related to indexing and phinx.

Not sure if you would be interested in merging this branch into yours, or if you would need more from my end before doing so.

Either way, thanks for the effort putting this schema dumper together, it has saved me several hours worth of work.

### CODE REVIEW

**Comment by basilfx** [comment]
Sounds like a great addition! Let me take a look at it.

File: src/Phinx/Db/Adapter/AdapterInterface.php [hunk 1]
```diff
@@ -47,6 +47,7 @@
     const PHINX_TYPE_CHAR           = 'char';
     const PHINX_TYPE_TEXT           = 'text';
     const PHINX_TYPE_INTEGER        = 'integer';
+    const PHINX_TYPE_DOUBLE         = 'double';
```
Comment (basilfx) [inline src/Phinx/Db/Adapter/AdapterInterface.php]: I think this would be better in a separate PR.

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 2]
```diff
@@ -857,6 +864,9 @@ public function getSqlType($type, $limit = null)
             case static::PHINX_TYPE_FLOAT:
                 return array('name' => 'float');
                 break;
+            case static::PHINX_TYPE_DOUBLE:
```
Comment (basilfx) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: I think this would be better in a separate PR.

File: src/Phinx/Db/Adapter/MysqlAdapter.php [hunk 3]
```diff
@@ -533,17 +534,22 @@ public function dropColumn($tableName, $columnName)
      * Get an array of indexes from a particular table.
      *
      * @param string $tableName Table Name
+     * @param boolean $lowercase
      * @return array
      */
-    protected function getIndexes($tableName)
+    public function getIndexes($tableName, $lowercase = true)
```
Comment (basilfx) [inline src/Phinx/Db/Adapter/MysqlAdapter.php]: What's the added benefit of lowercase?

**Review by basilfx — CHANGES_REQUESTED** [review]

**Comment by basilfx** [comment]
Can you also rebase (interactively) to fixup 2d8ebb3 in 8904d29?

File: src/Phinx/Db/Adapter/MysqlAdapter.php (hunk 3 above)
Comment (jwh315) [reply src/Phinx/Db/Adapter/MysqlAdapter.php]: this method was originally lowercasing all index column names (as it is currently).  I initially thought I needed to have case sensitive column names for mysql on some operating systems.  That turned out to be incorrect thinking on my part.  I have removed the extra param

**Comment by jwh315** [comment]
I removed the DOUBLE column type and fixed the commit history to remove the superfluous commit.  I also noticed another issue that needed to be addressed for tables that do not have primary keys, that has been fixed in the most recent commit.

**Comment by basilfx** [comment]
Looks good to me, thanks!

**Comment by basilfx** [comment]
I made a [few improvements](https://github.com/basilfx/phinx/commit/1bb96f37fa1a9164d7289c4e2969f3f1221610e3) to the code:
- There was still a reference to double in the PDOAdapter
- Undefined index check
- Optimized template

Other than that, it works great :-)

File: src/Phinx/Migration/Schema/Schema.template.php [hunk 4]
```diff
@@ -15,6 +15,12 @@ public function change()
                 ->addColumn(<?php echo CodeGenerator::buildAddColumnArgumentsString($column);?>)
             <?php endif; ?>
         <?php endforeach; ?>
+
+        <?php foreach($table->getIndexes() as $name => $index): ?>
+            <? if ($name !== 'PRIMARY'): ?>
```
Comment (futoase) [inline src/Phinx/Migration/Schema/Schema.template.php:20]:
``` php
<?php if ($name !== 'PRIMARY'): ?>
```

Comment (basilfx) [reply src/Phinx/Migration/Schema/Schema.template.php:20]: Fixed, thanks!

**Comment by basilfx** [comment]
@jwh315 I have fixed some more:
- 84cc9b7d48ba5e4a377815ccb80befebbbc00ee8 reverts another change that should not be part of this fork.
- 926918f64559eee8cf2be42a0c99bcf8006541e0 column names should not be converted to lower case, just interpret them as returned by MySQL's `SHOW INDEXES`.
- 322eedf9e9affa0551a9d543456065e2f2daab4d fixes a problem with the code generation: in `buildIndexString` the `$name` parameter caused issues. In SQLite, the name of an index cannot match a column name. I remove this parameter, but added it back as an option of the `Index` class.

**Pull request merged**